### ISSUE TITLE
Field instance config files are not being deleted when fully uninstalling Comment

### ISSUE THREAD

**Issue by indigoxela** [backdrop/backdrop-issues#5469]
## Description of the bug

The comment module does an incomplete cleanup, when uninstalled. This leads to PHP Notices on the config export page.

## Steps To Reproduce

1. Turn on error display
2. Uninstall the comment module
3. Go to /admin/config/development/configuration/single/export

## Actual behavior

Two PHP notices:

    Notice: Trying to access array offset on value of type null in field_config_label_instance() (Zeile 435 von .../core/modules/field/field.module).
    Notice: Trying to access array offset on value of type null in field_config_label_instance() (Zeile 435 von .../core/modules/field/field.module).

## Expected behavior

No notices, the comment module should remove all related config.

## Additional information

When the comment module gets uninstalled it does run `field_delete_field('comment_body');`, but two files stay untouched:

- field.instance.comment.comment_node_page.comment_body.json
- field.instance.comment.comment_node_post.comment_body.json

And these two files cause the php notices on the "Single export" configuration manager page. The Comment entity no longer exists, there's no instance label.

It may be worth checking, if this incomplete uninstall also affects other modules, for instance Taxonomy. Update: no, the Taxonomy module is not affected.

- Backdrop CMS version: latest stable, 1.21.1
- PHP version: 7.4

**Comment by argiepiano** [comment]
I have been able to reproduce this issue with fieldable custom entities as well. 

1. I installed the module that creates a fieldable custom entity
2. I added a new Field API field. Config files for both the field info (`field.field.[FIELD_NAME].json`) and field instance (`field.instance.[ENTITY_TYPE].[FIELD_NAME].json`) are created
3. I created and saved an instance of the entity through the UI and then deleted it.
4. I uninstalled the module that created the custom entity. `field_delete_field()` was called at this time
5. When checking the config folder, I noticed that the field info configuration file (`field.field.[FIELD_NAME].jason`) was removed (as it should), but the field instance configuration file was not
6. Then, when visiting  /admin/config/development/configuration/single/export I get the same notice you reported above

**Comment by klonos** [comment]
Would it then make sense to iterate through all entities, find all fields added to comments, and make sure that the instances that are not in use are deleted before completely uninstalling the Comment module? (not when disabling it I guess, cause people might reenable it).

PS: We are already warning about data loss when uninstalling the module:

![image](https://user-images.githubusercontent.com/2423362/150653993-862bce2c-d0e5-4895-b651-118d78895d90.png)

**Comment by indigoxela** [comment]
I belief, the problem arises further down the road.

When uninstalling Comment, the module is already disabled, hence the Comment entity info is no longer available, which seems to influence the instance deletion. Might need a little more digging.

**Comment by klonos** [comment]
> ...it's only the instance config file that stays there.

That kinda rings a bell. I think that we had some issue re removing config files that are no longer owned by any active module. Couldn't find it though.

**Comment by echozone** [comment]
I get same Notice plus the second Notice, on config manger single export screen, but comment module is installed + I did not uninstall it. It's a D7 migrated site.

`Notice: Trying to access array offset on value of type null in field_config_label_instance() (line 435 of __/core/modules/field/field.module).
Notice: Trying to access array offset on value of type null in field_config_label_bundle() (line 446 of __/core/modules/field/field.module).`

**Comment by argiepiano** [comment]
This bug appears ALSO when the Comment module is disabled but not fully uninstalled.

I believe one should not be able to export a field instance config file when the entity the field is attached to is not available. In this case, when Comment is disabled or uninstalled. Is this correct? 

Additionally, we don't want to delete the field info and the instance info CMI files until the module is completely uninstalled, right? This gives the user the chance of re-enabling and having the data still intact.

If the answer is yes to both of the above, then the solution is to skip including the comment_body and other fields attached to comment within `config_get_prefix_groups()` if the entity definition is not available.  This is a bit tricky to do, as `config_get_prefix_groups()` is really agnostic about the type of config file it's dealing with. So, this checking would only apply to config prefixes `field.instance`.

Do you think this is the right approach?

**Comment by indigoxela** [comment]
To prevent an export of a field instance config, when the entity the field is attached to is not available, might solve at least a part of the problem.

The base problem - on *uninstall* config that should get deleted, doesn't get deleted - would still be unsolved, though. But let's fix what's fixable. If it is.

**Comment by argiepiano** [comment]
> To prevent an export of a field instance config, when the entity the field is attached to is not available, might solve at least a part of the problem.
> 
> The base problem - on _uninstall_ config that should get deleted, doesn't get deleted - would still be unsolved, though. But let's fix what's fixable. If it is.

Yes. It seems like we should split this issue into two:  (1) one dealing with the fact that config export  functionality is trying to access an unavailable entity/bundle (entities created by "disabled-only" modules, where the config file should not be deleted), and (2) another dealing with the fact that completely uninstalling comment doesn't remove field instance config files attached to comment entities. 

How about we rename this to fit the first case and open a second issue to fit the second case?

**Comment by indigoxela** [comment]
> How about we rename this to fit the first case and open a second issue to fit the second case?

Hm, this issue's description is clearly about the second case - the title isn't that specific.

How about opening a new issue for the config export problem and making this issue's title more specific, so it's clearer what it's about?

**Comment by argiepiano** [comment]
That's OK with me. 

I would still like to tackle the problem of the field instance config files not being deleted when completely uninstalling Comment. I'll open a separate issue for that.

**Comment by indigoxela** [comment]
> tackle the problem of the field instance config files not being deleted when completely uninstalling Comment. I'll open a separate issue for that.

This IS the issue for that :grinning: [This comment](https://github.com/backdrop/backdrop-issues/issues/5469#issuecomment-1019443466) tries to explain why this happens. But there's no (clean) solution for that yet. It also affects other modules, btw.

**Comment by argiepiano** [comment]
Well, if this is the issue for that, then the title should clearly indicate that. Something like

"Field instance config files are not being deleted when fully uninstalling Comment"

**Comment by indigoxela** [comment]
Issue title changed as suggested.

**Comment by yorkshire-pudding** [comment]
I can reproduce this issue.  Also the one where it is just disabled.

**Comment by indigoxela** [comment]
The bug's still present, btw. :wink:

**Comment by alanmels** [comment]
One of our project never uses comment_body field, so it's deleted for all content types and we hit the reported error. Anything deleted should properly clean up after itself, however I believe it wouldn't hurt also to introduce an additional check at core level. In particular, the `field_config_label_instance()` function that now looks like:

```php
function field_config_label_instance($instance, $config_name) {
  list($entity_type_name, $bundle, $field_name) = explode('.', str_replace('field.instance.', '', $config_name));
  $entity_type = entity_get_info($entity_type_name);
  $entity_label = $entity_type['label'];
  $bundle_label = isset($entity_type['bundles'][$bundle]['label']) ? $entity_type['bundles'][$bundle]['label'] : $bundle;
  return $entity_label . ' - ' . $bundle_label . ' - ' . $instance['label'];
}
```

could be rewritten as:

```php
function field_config_label_instance($instance, $config_name) {
  list($entity_type_name, $bundle, $field_name) = explode('.', str_replace('field.instance.', '', $config_name));
  $entity_type = entity_get_info($entity_type_name);
  if (!isset($entity_type['label'])) {
    $entity_label = $entity_type['label'];
    $bundle_label = isset($entity_type['bundles'][$bundle]['label']) ? $entity_type['bundles'][$bundle]['label'] : $bundle;
  return $entity_label . ' - ' . $bundle_label . ' - ' . $instance['label'];
  }
}
```

to escape the warning message. Or make the core give some useful information like:

```php
function field_config_label_instance($instance, $config_name) {
  list($entity_type_name, $bundle, $field_name) = explode('.', str_replace('field.instance.', '', $config_name));
  $entity_type = entity_get_info($entity_type_name);
  if (!isset($entity_type['label'])) {
    backdrop_set_message(t('A configuration file for the deleted field @field_name is found. You might consider deleting it.', array('@field_name' => $config_name)), 'warning');
  }
  else {
    $entity_label = $entity_type['label'];
    $bundle_label = isset($entity_type['bundles'][$bundle]['label']) ? $entity_type['bundles'][$bundle]['label'] : $bundle;
    return $entity_label . ' - ' . $bundle_label . ' - ' . $instance['label'];
  }
}
```

**Comment by indigoxela** [comment]
@alanmels many thanks for giving this a try. :+1: 

Your PR contains syntax errors, though. And also some coding standard violation.
The automated comments should guide you to fix that: https://github.com/backdrop/backdrop/pull/4876/files

**Comment by alanmels** [comment]
> @alanmels many thanks for giving this a try. 👍
> 
> Your PR contains syntax errors, though. And also some coding standard violation. The automated comments should guide you to fix that: https://github.com/backdrop/backdrop/pull/4876/files

Sorry, was too much in a hurry yesterday, didn't run proper tests. I'm not sure if it's an ideal solution, though. Just wanted to bring more attention to this issue.

**Comment by indigoxela** [comment]
> I'm not sure if it's an ideal solution, though. Just wanted to bring more attention to this issue.

It's an annoying bug. Any suggestion to better handle that is helpful. :+1:

**Comment by argiepiano** [comment]
This is a complex problem. To help us strategize a solution, in a nutshell this is what's happening:

1. After you disable `comment`, if you run `field_info_field('comment_body')` to get info about the field (which still exists, as the module has not been uninstalled), you get an array with an empty `bundles` subarray. This is because, rightly so, the field is not currently attached to any bundle of the `comment` entity. 
2. Then, when you uninstall `comment`, Backdrop invokes `comment_uninstall()`, which in turns invokes `field_delete_field('comment_body')`. 
3. `field_delete_field()` normally does two things: it removes all existing instance definitions for a field, and then it removes the field definition. To remove instance definitions, the function uses the `bundles` subarray of the info array returned by `field_info_field`. 
4. In the case of a disabled module like `comment`, this `bundles` subarray is empty (see number 1 above).
5. The result is that the instance definitions (json files) are never deleted in this case

This, by the way, will most likely also affect other fieldable entities that can be disabled (e.g. custom entities like Ubercart's and Registration's entities).

So, I see three possible solutions
1. Quick and dirty, solves the problem for `comment`: in `comment_uninstall` we could load all instance definition json files (ie those that start with `field.instance.comment.`) and remove them manually. Maintainers of modules that define entities should then follow this pattern.
2. Or, we could modify `field_delete_field()` to check if both conditions are true: the info field array has an empty `bundles` subarray, AND there are existing field instance json files for that given entity. If so, then invoke `field_instance_instance` as it would otherwise.

I'm going to attempt the second solution in a PR.

**Comment by alanmels** [comment]
@argiepiano, thanks for your input. Please note that the Comment module may still be enabled across the website, but the `comment_body` field might have been deleted for one or more content types. In my opinion, the issue isn't related to the status of the Comment module being disabled or uninstalled, but rather ensuring the proper deletion of the `comment_body` field and tidying up the corresponding configuration file.

**Comment by argiepiano** [comment]
> In my opinion, the issue isn't related to the status of the Comment module being disabled or uninstalled, but rather ensuring the proper deletion of the comment_body field and tidying up the corresponding configuration file.

@alanmels I'm not sure this is the case. In fact, this is easy to test:
1. Go to each of your content types and click on the "Comment fields"
2. Remove the Body (`comment_body`) field from each and all of them

Then check your config folder: all json files related to `comment_body` (both field info and instance info) are removed normally.

**Comment by alanmels** [comment]
@argiepiano, thanks for the further details. On the project we had hit this issue the comment module was enabled, but the `comment_body` field was deleted. However, it might be possible that at some point the comment module was disabled and then re-enabled. Anyway, I trust your assessment and do hope this will be somehow addressed as the current state is not ideal.

**Comment by indigoxela** [comment]
> This, by the way, will most likely also affect other fieldable entities that can be disabled

Yes, that's likely - with the current field deletion logic.

Personally, I'd also be happy with a workaround like the proposed one in this PR - warn people, give a helpful hint, if abandoned field conf is determined.
Anything seems better than the current situation, which is just an obscure PHP warnings on an admin page.

A clean solution for all entities would be ideal, but that appears to be a bit tricky. Given, this issue is open now for two years and no easy solution available, yet - I'd also go with a workaround.

**Comment by argiepiano** [comment]
One more data point: D7 suffers from the same issue. When disabling `comment`, all field data and field and instance infos remain in the database. This is not directly relevant to Backdrop, but it's evidence that this issue was never addressed properly in D7.

**Comment by argiepiano** [comment]
I have submitted a PR that fixes the issue from its root cause. The PR adds code to the hook implementation `entity_modules_uninstalled()` to remove leftover field info and field instance json files.  Notice that at this point, using function `field_info_field()` and `field_info_instance()` will not return any data related to the entities defined in the module being uninstalled. This is because the module was initially disabled before being uninstalled, and therefore (rightly so) these functions will not include fields attached to the entity that was disabled previously.  Therefore, the PR actually has to deal with leftover json files directly.

To test:
1. Better to test on a clean install of Backdrop
2. For complete testing, add a new dummy text field  called "A text" to both: the Post content type, AND to the comments for the Post content type. 
3. Add a Post and some comments
4. Check the active config folder to verify that the `field.field.comment_body.json`, `field.field.field_a_text.json`, and all its related instance files with the pattern  `field.instance.comment.X` exist
5. Disable `comment`
6. Check the above files again. They should still exist
7. Uninstall `comment`
8. Check the files above. All files EXCEPT  `field.field.field_a_text.json` should have been removed.

Notice also that I remove `field_delete_field()` from `comment.uninstall`, since this is really not needed.

NB: All field and entity table removal is done by other functions and do not need to be addressed in this PR.

**Comment by indigoxela** [comment]
Oh, that works a treat! :+1: 

Some things to consider, though:

If the comment module's only _disabled_, but not uninstalled, yet, the warnings re `Warning: Trying to access array offset on null in field_config_label_instance()` are still thrown on admin/config/development/configuration/single/export
They go away, as soon as comment's uninstalled.
I wonder, if we could handle that more gracefully on the export page. That might be a follow-up, or maybe we can fix it here? Maybe (sort of) combine both approaches of the current PR?

After uninstalling comment module, the field values of the deleted field instance are still in the database. In my case it was a number field "field_dummy". Also after running cron.
Three columns in the db tables indicate, that it's a zombie: entity_type = comment, bundle = comment_node_post, deleted = 1.
I wonder, why these rows don't go away after cron run - maybe I'm misunderstanding the logic behind the "deleted" column? Might be a follow-up as well, to dig into that.

Currently the PR fixes things, when uninstalling comment module with this patch applied. That means, when the comment module has been uninstalled before this change, the problem persists (hook_modules_uninstalled won't run again).
Should we catch that in an update hook? And if it were only for the comment module, which is the most likely candidate for this nagging (it's on by default on fresh installs).

**Comment by indigoxela** [comment]
I think, the remaining rows should get handled in a follow-up. field_purge_batch() seems to be based on field_read_instances() - this can't work for deleting rows of uninstalled modules that shipped with entities, that share common fields with entity types still active. Sort of an edge case, it seems, and unrelated to the problem we try to solve here, anyway.

**Comment by argiepiano** [comment]
Thanks for taking a look, @indigoxela!

>If the comment module's only disabled, but not uninstalled, yet, the warnings re Warning: Trying to access array offset on null in field_config_label_instance() are still thrown ... I wonder, if we could handle that more gracefully on the export page. That might be a follow-up, or maybe we can fix it here? Maybe (sort of) combine both approaches of the current PRs?

Makes sense. I can try that.

> After uninstalling comment module, the field values of the deleted field instance are still in the database.

Oh, this is strange. I thought I had checked that and those were removed. I will try this again. 

> That means, when the comment module has been uninstalled before this change, the problem persists (hook_modules_uninstalled won't run again). Should we catch that in an update hook?

Good point. I think an update hook is the way to go, but I'd like for this to really work beyond the comment module. Let me try this in the next couple of day.

**Comment by indigoxela** [comment]
@argiepiano that was my fault, sorry for the noise re field data when the field is only attached to the unavailable entity.
That works as expected, as the table gets completely dropped.
Only the "mixed entity" field data remains in db (with the column "deleted" set to 1). Not deletable by cron, though. We still could make that a follow-up, if you prefer.

**Comment by argiepiano** [comment]
> Only the "mixed entity" field data remains in db (with the column "deleted" set to 1). Not deletable by cron, though. We still could make that a follow-up, if you prefer.

Thank you! Great catch. Let me come back to this later to investigate a way to take care of this. I think we should try to do this as part of this issue as well.

**Comment by argiepiano** [comment]
I have revised my PR. @indigoxela and @alanmels, I'll appreciate your testing and thoughts about this

First, a summary of the current behavior. This will help people understand my approach in the PR.

<details>
<summary>Current behavior when uninstalling comment</summary>

To demonstrate the current behavior, in addition to the `comment_body` default field, I added two fields:
- `field_a_shared_field`: this is a text field added to both, the `post` node bundle, and to the `comment_node_post` bundle (thus this field is used both in `comment` and `node` entities)
- `field_a_unique_field`: this is a text field added ONLY to the `comment_node_post` comment bundle

I added some sample `post` content and  some comments to that post, before uninstalling `comment`

After fully uninstalling `comment`, this is what's happening (before the PR):

**For the field `field_a_shared_field`:**
1.  Database records for this field as marked as deleted (i.e. the `deleted` column is set to 1) 
2. The field info json `field.field.field_a_shared_field.json` is NOT removed 👍🏽 
3. The instance json `field.instance.comment_node_post.field_a_shared_field.json`  is NOT removed but is marked as `deleted` 👎🏽  (see below why this is bad)

**For the field `field_a_unique_field`:**
1.  The database tables `field_data_field_a_unique_field` and `field_revision_field_a_unique_field` are completely removed 👍🏽 
2. The field info json `field.field.field_a_unique_field.json` IS removed 👍🏽 
3. The instance json `field.instance.comment_node_post.field_a_unique_field.json` is NOT removed , but is marked as `deleted` 👎🏽 

**For the field `comment_body`:**
1.  The database tables `field_data_comment_body` and `field_revision_comment_body` are completely removed 👍🏽 
2. The field info json `field.field.comment_body.json` IS removed 👍🏽 
3. The instance json `field.instance.comment_node_post.comment_body.json` is NOT removed , and is NOT marked as `deleted` 👎🏽 
4. The instance json `field.instance.comment_node_page.comment_body.json` is NOT removed , and is NOT marked as `deleted` 👎🏽 
5. The instance json `field.instance.comment_node_card.comment_body.json` is NOT removed , and IS marked as `deleted` (???) 👎🏽 

Then, clearing the caches and running cron will not remove either the field data that remained (see `field_a_shared_field`), nor the remaining instances from the 3 fields.  👎🏽 

Additionally, since the `comment` entity type doesn't exist anymore, **nothing** will work to remove those remaining data, even though they are marked as `deleted`
</details>

The PR is simple but it may raise some eyebrows. At first sight, the PR seems a bit unorthodox, as it directly deletes the remaining instance json files, and removes the remaining database records for fields that are shared among `comment` entities and other entities. 

The reason for this approach is that nothing else works. In theory, `field_cron()` should be removing data and instances marked as `deleted`, but since the `comment` entity type doesn't exist anymore (the module is uninstalled) `field_cron()` fails to take action on these "zombie" data (see below for an explanation of why).  At some point, perhaps it'd be a good idea to refactor `_field_invoke_get_instances()` or related functions to include  "zombie" instances of fields attached to entity types that don't exist anymore. But at this point, this is the safest approach, and the best I can do.

<details>
<summary>Why `field_cron()` doesn't work with these fields</summary>

Basically, this is the sequence of events:

1. `field_cron()` invokes `field_purge_batch()`
2. ... which invokes `field_purge_data()`
3. ... which invokes `_field_invoke('delete'...`
4. ... which invokes `_field_invoke_get_instances()`
5. ... which returns `null` for field instances attached to inexistent entity types like `comment`
</details>

**Comment by argiepiano** [comment]
I want to add that those "zombie" entries in the shared field table may seem harmless, but they are not. They have the potential to produce a fatal error when you: (1) add comments to a post, (2) unsintall `comment`, (3) re-install comment, (4) add comments to a post, (5) uninstall `comment` a second time (fatal error, duplicate record in the database)

**Comment by indigoxela** [comment]
@argiepiano many thanks for this great summary of the problem we're facing here. :+1: 
I'll try to do more testing with the latest changes soon.

**Comment by robertgarrigos** [comment]
I just came across this issue, as well as #5350. Is this issue fixing the fact that other unused field.instance config files are not being deleted when a content type is deleted (not just the comment field instance file)?
